wurst/extensions/UnitExtensions.wurst

```diff
@@ -134,19 +134,45 @@ public function unit.setCreepGuard(bool value)
 
 public function unit.addOrStackItem(item itm) returns bool
     //Stackables
+    var result = false
+    let itemPos = itm.getPos()
     if isItemStackable(itm.getTypeId())
-        if not this.isItemFullyStacked(itm.getTypeId()) and this.hasItemById(itm.getTypeId())
-            this.stackItem(itm)
-            return true
-        else if not this.isInventoryFull()
-            this.addItemHandle(itm)
-            return true
+        this.stackItem(itm)
+        if itm.isAlive()
+            //Add leftover stacks if they were left
+            var charges = itm.getCharges()
+            let itemTypeId = itm.getTypeId()
+            let maxStack = getItemMaxStacks(itemTypeId)
+
+            while charges > 0
+                if charges > maxStack
+                    if not this.isInventoryFull()
+                        let newItemStack = createItem(itemTypeId, itemPos)
+                        newItemStack.setCharges(maxStack)
+                        this.addItemHandle(newItemStack)
+                        charges -= maxStack
+                        itm.setCharges(charges)
+                        result = true
+                    else
+                        result = false
+                        break
+                else
+                    itm.setCharges(charges)

```

Comment:
Redundant assignment. Whenever local variable `charges` is modified, you reflect the change on the itm handle immediately.